download_plugins.py

```diff
@@ -1,85 +1,160 @@
+            )
+            get_dependencies(dependency["name"])
+
+
+# Download plugin from jenkins update server.
+# If no version is specified, latest plugin is downloaded.
+def download_plugin(plugin, version=None):
     download_url = ""
-    
+
     if version is None:
-        download_url = plugin_base_url + "/latest/%s.hpi" % plugin
+        download_url = "{url}/latest/{plugin}.hpi".format(
+            url=plugin_base_url, plugin=plugin
+        )
     else:
-        download_url = plugin_base_url + "/download/plugins/%(plugin)s/%(version)s/%(plugin)s.hpi" % {'plugin': plugin, 'version': version}
+        download_url = (plugin_base_url +
+                        "/download/plugins/" +
+                        "{plugin}/{version}/{plugin}.hpi".format(
+                            plugin=plugin, version=version
+                        ))
 
     plugin_download = requests.get(download_url, stream=True)
 
-    if plugin_download.status_code is 200:
-        destination_path = download_directory + "/" + plugin + ".hpi"
+    if plugin_download.status_code == 200:
+        destination_path = "{dir}/{plugin}.hpi".format(
+            dir=download_directory, plugin=plugin
+        )
 
-        with open(destination_path, 'wb') as data:
+        with open(destination_path, "wb") as data:
             for chunk in plugin_download.iter_content(chunk_size=128):
                 data.write(chunk)
 
-        print "Downloaded", plugin
-        
+        print("Downloaded {}".format(plugin))
+
     else:
-        print "Error downloading %s. Response:" % plugin
-        print plugin_download.text
+        print("Error downloading {plugin}. Response:\n{response}".format(
+            plugin=plugin, response=plugin_download.text
+        ))
+        global exit_code
         exit_code = 1
 
-    installed_plugins.append(plugin)
 
 # Install each plugin in the supplied json file along with dependencies.
 def install_plugins():
     for plugin, version in plugins.items():
-        print "**** Install Plugin: %s ****" % plugin
+        print("\n**** Install Plugin: {} *****".format(plugin))
         download_plugin(plugin, version)
         get_dependencies(plugin)
 
+    print("\nDownloading all dependencies")
+    print("*****************************************")
+    for plugin, version in stored_plugins.items():
+        download_plugin(plugin, version)
+
+    for dependency, info in dependencies_info.items():
+        if not (info["duplicate"]):
+            continue
+        print(dedent("""\
+        Warning: Multiple versions of {} found in dependencies.\
+        """.format(dependency)))
+        for parent in info["parents"]:
+            for version, plugin in parent.items():
+                print("    Version {version} required by {plugin}".format(
+                    version=version, plugin=plugin
+                ))
+        print("    Downloaded version {}".format(stored_plugins[dependency]))
+
```

Comment:
Like we discussed in person, could you add a logic to handle if the version of the top-level plugins if higher than the highest dependencies? Also, if the version of the top-level plugins is lower than downloaded dependencies version, then we might also want to download the higher one. However, we should also print the warning message for this scenario. For example, when I used this script to download this list of plugins:

```yaml
plugins:
  ant: "1.8"
  build-name-setter: "1.6.9"
  build-user-vars-plugin: "1.5"
  cas-plugin: "1.4.1"
  cloudbees-folder: "6.4"
  config-file-provider: "2.18"
  copyartifact: "1.39"
  credentials-binding: "1.16"
  dashboard-view: "2.9.11"
  email-ext: "2.62"
  envinject: "2.1.5"
  environment-script: "1.2.5"
  file-leak-detector: "1.4"
  git: "3.8.0"
  github: "1.29.0"
  git-client: "2.7.1"
  gradle: "1.28"
  hipchat: "2.1.1"
  jdk-tool: "1.1"
  ldap: "1.20"
  log-parser: "2.0"
  mailer: "1.20"
  matrix-auth: "2.2"
  maven-plugin: "3.1"
  parameterized-trigger: "2.35.2"
  publish-over-ssh: "1.19.1"
  rebuild: "1.27"
  role-strategy: "2.7.0"
  ssh-agent: "1.15"
  ssh-slaves: "1.26"
  windows-slaves: "1.3.1"
  workflow-aggregator: "2.5" #pipeline
  workflow-cps: "2.45"
  workflow-job: "2.17"
  ws-cleanup: "0.34"
```

I got the following message:

```
Warning: Multiple versions of cloudbees-folder found in dependencies.
    Version 5.12 required by workflow-cps-global-lib
    Version 6.1.0 required by branch-api
    Version 6.1.2 required by workflow-multibranch
    Downloaded version 6.1.2
```

It seems not correct since we need to download `cloudbees-folder: "6.4"` as we defined in the top-level plugins list.